MAST refactor: observations.py

ceb8 opened the pull request:
In this PR I moved the ObservationsClass into observations.py and refactored it to use the new PortalAPI class instead of being a subclass of the old MastClass.

I left the original ObservationsClass code in core, but renamed it ObservationsClass_old (it will get deleted when all functionality is out of core.py and I remove that file).

tomdonaldson commented on astroquery/mast/observations.py:
```diff
@@ -0,0 +1,902 @@
+    Class for querying MAST observational data.
+    """
+
+    def __init__(self, mast_token=None):
```
I'm wondering whether `__init__()`, `_login()`, `get_token()`, `session_info()`, `logout()` or `_parse_result()` are any different than in `MastClass`?  If not (and if I'm reading it right), then that seems like too much code to duplicate.  I can see a couple options and am open to others.

1)  Extract a common base class that includes those methods (or whatever the actual common behavior is).  Maybe call it something like `MastQueryWithLogin`?

2) Go back to having `ObservationClass` extend `MastClass`.  I know this defeats some of the purpose of this refactoring, but since `MastClass.resolve_object()` is just a pass through to `utils` (and maybe not needed?), `MastClass.service_request_async()` is the only thing `MastClass` has that `ObservationClass` doesn't.

ceb8 replied:
To a certain extent I think we have to live with more code reuse than we had before in order to make the code more extendable. And some of it also comes from the goal of leaving the interface untouched from a user perspective.

The `utils.resolve_object` for instance.  Ultimately I think what we want to do is deprecate it from all the classes and use only the utils version, but we can't just get rid of it. Last time I messed with that function it broke lightkurve, so we'll need to have an actual deprecation period and probably open an issue over at lightkurve (and maybe eleanor) to make sure that doesn't happen again.

In terms of the massive overlap between `MastClass` and `ObservationsClass`. I don't want to make `ObservationsClass` a subclass of `MastClass` because the point of separating out the portal connection was so that `ObservationsClass` can also have microservice connections in future. Something that I did debate though was whether to separate `MastClass` from the `PortalAPI` class at all. Now might be a time to revisit that idea.  I could move the `MastClass` functionality into `PortalAPI` and then just define `MastClass = PortalAPI` and `Mast = PortalApi()`. It wouldn't completely eliminate the redundancy, but it would lower it. 

Another driver of the code reuse is that I've based all of the classes on the Astroquery base classes. I did this because I wanted to use the functionality you get from that, and be consistent with the Astroquery design overall, but it does also mean we need to have some of the same functions in all the classes. I'd be interested in any opinions on this as a design choice.

What do you and @phicc  think?

tomdonaldson replied:
I agree that it makes sense for `ObservationClass` not to inherit from `MastClass` for the reason you state.  It does seems like too much duplicate code to me though, so my other thought was to extract the commonality into a new class from which they both inherit.  

That approach makes the most sense if that commonality is expected to persist once `ObservationClass` starts getting the extensions you mention and if the commonality neatly fits a single concept or functionality.  In this case I was wondering if that commonality was the login capability, in which case it might make sense to add a class that inherits from `QueryWithLogin` and adds MAST login capabilities.  It could then be extended by both `ObservationClass` and `MastClass`.

tomdonaldson commented on astroquery/mast/observations.py:
```diff
@@ -0,0 +1,902 @@
+        self._auth_obj.logout()
+        self._authenticated = False
+
+    def _parse_result(self, responses, verbose=False):
```
For `_parse_result()` both here and in `MastClass`, is it called anywhere and/or is it needed?

ceb8 replied:
It's part of the @async_to_sync functionality. When you mark a function as async the decorator causes the class to make a new function that isn't async which is the output of the async function run through _parse_result.

tomdonaldson replied:
OK, that makes sense.  The fact that that wasn't obvious to me is a good example of why I'm not enamored with decorators.  But I have to admit that this seems like a good case for one since that pattern is repeated over and over again.

tomdonaldson commented on astroquery/mast/observations.py:
```diff
@@ -0,0 +1,902 @@
+
+    def __init__(self, mast_token=None):
+
+        super(ObservationsClass, self).__init__()
```
Not something for now since it would be nice to keep consistency, at least within the mast module, but I wonder if we should start using the Python 3 style that I think would just be `super().__init__()`.

ceb8 replied:
Oh I didn't know that had changed. I will look into it and upgrade. I'm trying to Python 3-ify as I go along.

ceb8 commented:
@tomdonaldson @phicc I created a new base class for astroquery.mast called `MastQueryWithLogin` that extends `QueryWithLogin` and does all the setup/login/cloud initialization that I expect to be consistent across MAST classes. I also added the `resolve_object` function to it. This should decrease the code duplication while maintaining our flexibility.

ceb8 commented:
@tomdonaldson @phicc I went ahead and made the `ServiceAPI` class, refactored `CatalogsClass` and removed all the deprecated code, so it's easier to see what's going on. I think it's in pretty good shape at this point. `ServiceAPI` is  too specific to the PanSTARRS catalog, but I don't think it's terribly useful to fix that without having another reference point. My plan is to start that process when I refactor the TESScut code, it should be able to use `ServiceAPI` and will provide a good example for what needs to go into that class and what should be moved into `CatalogsClass`.

I've got all the tests working as well, however I am getting a lot of timeouts in the remote TESScut tests that I don't understand (I'm not getting any emails about TESScut being down), so I'll keep working on that.
